ndcube/meta.py

```diff
@@ -0,0 +1,249 @@
+import copy
+import numbers
+import collections.abc
+
+import numpy as np
+
+__all__ = ["Meta"]
+
+
+class Meta(dict):
+    """
+    A sliceable object for storing metadata.
+
+    Metadata can be linked to a data axis which causes it to be sliced when the
+    standard Python numeric slicing API is applied to the object.
+    Specific pieces of metadata can be obtain using the dict-like str slicing API.
+    Metadata associated with an axis/axes must have the same length/shape as those axes.
+
+    Parameters
+    ----------
+    header: dict-like
+        The names and values of metadata.
+
+    comments: dict-like (optional)
+        Comments associated with any of the above pieces of metadata.
+
+    axes: dict-like (optional)
+        The axis/axes associated with the above metadata values.
+        Each axis value must be None (for no axis association), and `int`
+        or an iterable of `int` if the metadata is associated with multiple axes.
+        Metadata in header without a corresponding entry here are assumed to not
+        be associated with an axis.
+
+    data_shape: iterable of `int` (optional)
```

Comment:
```suggestion
    data_shape: `iterable` of `int`, optional
```
Not sure if this links.